lib/pengine/rules.c

```diff
@@ -443,74 +443,106 @@ phase_of_the_moon(crm_time_t * now)
-        }                                                               \
-	if(pass == FALSE) {						\
-	    crm_debug("Condition '%s' in %s: failed", value, xml_field); \
-	    return pass;						\
-	}								\
-	crm_debug("Condition '%s' in %s: passed", value, xml_field);	\
+static pe_eval_date_result_t
+check_one(xmlNode *cron_spec, const char *xml_field, uint32_t time_field) {
+    pe_eval_date_result_t rc = pe_date_result_undetermined;
+    const char *value = crm_element_value(cron_spec, xml_field);
+    long long low, high;
+
+    if (value == NULL) {
+        /* Return pe_date_result_undetermined if the field is missing. */
+        goto bail;
     }
 
-gboolean
-pe_cron_range_satisfied(crm_time_t * now, xmlNode * cron_spec)
-{
-    const char *value = NULL;
+    if (pcmk__parse_ll_range(value, &low, &high) == pcmk_rc_unknown_format) {
+       goto bail;
+    } else if (low == -1 && high == -1) {
+        /* A single number was given, not a range. */
+       low = crm_parse_ll(value, NULL);
+        if (low > time_field) {
+            rc = pe_date_before_range;
+        } else if (low < time_field) {
+            rc = pe_date_after_range;
+        } else {
+            rc = pe_date_within_range;
+        }
+    } else if (low != -1 && high != -1) {
+        /* This is a range with both bounds. */
+        if (time_field < low) {
+            rc = pe_date_before_range;
+        } else if (time_field > high) {
+            rc = pe_date_after_range;
+        } else {
+            rc = pe_date_within_range;
+        }
+    } else if (low == -1) {
+       /* This is a range with no starting value. */
+        rc = time_field <= high ? pe_date_within_range : pe_date_after_range;
+    } else if (high == -1) {
+        /* This is a range with no ending value. */
+        rc = time_field >= low ? pe_date_within_range : pe_date_before_range;
+    }
 
-    long long value_low = -1;
-    long long value_high = -1;
+bail:
+    if (rc == pe_date_within_range) {
+        crm_debug("Condition '%s' in %s: passed", value, xml_field);
+    } else {
+        crm_debug("Condition '%s' in %s: failed", value, xml_field);
+    }
 
-    uint32_t h, m, s, y, d, w;
+    return rc;
+}
 
-    CRM_CHECK(now != NULL, return FALSE);
+static gboolean
+check_passes(pe_eval_date_result_t rc) {
+    /* _within_range and _op_satisfied are obvious.  _result_undetermined is a pass
+     * because this is the return value if a field is not given.  In this case, we
+     * just want to ignore it and check other fields to see if they place some
+     * restriction on what can pass.
+     */
+    return rc == pe_date_within_range || rc == pe_date_op_satisfied ||
+           rc == pe_date_result_undetermined;
+}
 
-    crm_time_get_timeofday(now, &h, &m, &s);
+#define CHECK_ONE(spec, name, var) { \
+    pe_eval_date_result_t subpart_rc = check_one(spec, name, var); \
+    if (check_passes(subpart_rc) == FALSE) { \
+        return subpart_rc; \
+    } \
+}
```

Comment:
Just as a general best practice, we wrap multi-line macros in do { ... } while (0). That construct can be used everywhere a single function call can, so it avoids non-obvious usage issues. E.g.

    if (a)
        CHECK_ONE(a, b, c);
    else
        return;

wouldn't compile because it would expand to if { ... }; else return; and the semi-colon would make the else hanging. Of course our style is to always use braces but it's still a good general practice.